telegram/ext/_conversationhandler.py

```diff
@@ -484,32 +525,43 @@ def map_to_parent(self, value: object) -> NoReturn:
             "You can not assign a new value to map_to_parent after initialization."
         )
 
-    @property
-    def persistence(self) -> Optional[BasePersistence]:
-        """The persistence class as provided by the :class:`Dispatcher`."""
-        return self._persistence
-
-    @persistence.setter
-    def persistence(self, persistence: BasePersistence) -> None:
-        self._persistence = persistence
-        # Set persistence for nested conversations
-        for handlers in self.states.values():
-            for handler in handlers:
-                if isinstance(handler, ConversationHandler):
-                    handler.persistence = self.persistence
+    async def _initialize_persistence(
+        self, application: 'Application'
+    ) -> TrackingDict[Tuple[int, ...], object]:
+        """Initializes the persistence for this handler. While this method is marked as protected,
+        we expect it to be called by the Application/parent conversations. It's just protected to
+        hide it from users.
 
-    @property
-    def conversations(self) -> ConversationDict:  # skipcq: PY-D0003
-        return self._conversations
+        Args:
+            application (:class:`telegram.ext.Application`): The application.
+
+        """
+        if not (self.persistent and self.name and application.persistence):
+            raise RuntimeError(
+                'This handler is not persistent, has no name or the application has no '
+                'persistence!'
+            )
+
+        with self._conversations_lock:
+            current_conversations = self._conversations
+            self._conversations = cast(
+                TrackingDict[Tuple[int, ...], object],
+                TrackingDict(),
+            )
+            # In the conversation already processed updates
+            self._conversations.update(current_conversations)
+            # above might be partly overridden but that's okay since we warn about that in
+            # add_handler
+            self._conversations.update_no_track(
+                await application.persistence.get_conversations(self.name)
+            )
 
-    @conversations.setter
-    def conversations(self, value: ConversationDict) -> None:
-        self._conversations = value
-        # Set conversations for nested conversations
-        for handlers in self.states.values():
-            for handler in handlers:
-                if isinstance(handler, ConversationHandler) and self.persistence and handler.name:
-                    handler.conversations = self.persistence.get_conversations(handler.name)
+        for handler in self._child_conversations:
+            await handler._initialize_persistence(  # pylint: disable=protected-access
+                application=application
+            )
+
+        return self._conversations
 
     def _get_key(self, update: Update) -> Tuple[int, ...]:
```

Comment:
in this function, we should also check if `chat is not None` right? Would also solve the mypy ignore ig

Reply:
I'll just have it throw `RuntimeErrors` when we get an invalid update. will never happen, but seems a bit cleaner to me.